rules/do-you-handle-multi-os-dev-teams-in-source-control/rule.md

```diff
@@ -0,0 +1,177 @@
+---
+seoDescription: Learn how to manage a multi-OS development team in source control. This article covers best practices for handling line endings (CRLF vs LF), creating cross-platform setup scripts, configuring Git for consistency, and using .NET CLI for secrets management to boost efficiency across Windows, macOS, and Linux.
+type: rule
+title: Do you handle Multi-OS dev teams in source control?
+uri: do-you-handle-multi-os-dev-teams-in-source-control
+authors:
+  - title: Tom Iwainski
+    url: https://www.ssw.com.au/people/thomas-iwainski
+related: []
+created: 2024-11-05T12:52:08.000Z
+archivedreason: null
+guid: 7bc2e9c8-2b12-466b-8ecd-34576e3a1648
+---
+
+In today’s development world, teams often consist of developers working on different operating systems (OS), such as Windows, macOS, and Linux. While each OS has its own strengths, managing a cross-platform development environment can introduce challenges. Issues like inconsistent line endings, platform-specific setup scripts, and configuration mismatches can lead to headaches for both individual developers and the team as a whole.
+
+<!--endintro-->
+
+## Addressing Line Ending differences (CRLF vs. LF)
+
+One of the most common issues faced by teams working with Git across different platforms is the handling of line endings. Windows uses **CRLF** (Carriage Return + Line Feed) for line endings, while macOS and Linux use **LF** (Line Feed) only. This can lead to unnecessary diffs in Git and potential merge conflicts.
+
+### Solution: Use Git’s Line Ending configuration
+
+Git provides a way to manage line endings across different operating systems by using the `core.autocrlf` setting. This configuration ensures that line endings are normalized when files are checked in and out of the repository.
+
+- **Windows Users**: Set Git to automatically convert line endings to **CRLF** when checking out files, and convert them back to **LF** when committing.
+
+  ```bash
+  git config --global core.autocrlf true
+  ```
+
+- **macOS/Linux Users**: Set Git to automatically convert **CRLF** line endings to **LF** when checking out files and leave them as **LF** when committing.
+
+  ```bash
+  git config --global core.autocrlf input
+  ```
+
+- **Repository-Wide Configuration**: It’s a good practice to enforce this configuration across the team via a `.gitattributes` file, which allows you to define how specific file types should be handled. For example:
+
+  ```text
+  * text=auto
+  *.cs text eol=lf
```

Comment:
I don't think this is right for `*.cs` files. We shouldn't have any specific configuration. I agree with the settings for `*.sh` and `*.bat`.

Reply:
even for *.cs files I am getting Visual Studio always popping up with the file ending has changed `do you want overwrite it?` and that is usually when our MacOS people touched a file :-D

I can take it out though, let me know

Reply:
I still don't think it's needed. I believe the problem is elsewhere.

Have a read of this post from Scott H: https://www.hanselman.com/blog/youre-just-another-carriage-return-line-feed-in-the-wall

Reply:
✅ Done - removed it.